core/src/banking_stage/transaction_scheduler/in_flight_tracker.rs

```diff
@@ -0,0 +1,142 @@
+use {
+    super::thread_aware_account_locks::ThreadId,
+    crate::banking_stage::scheduler_messages::TransactionBatchId, std::collections::HashMap,
+};
+
+/// Tracks the number of transactions that are in flight for each thread.
+pub struct InFlightTracker {
+    num_in_flight_per_thread: Vec<usize>,
+    cus_in_flight_per_thread: Vec<u64>,
+    batches: HashMap<TransactionBatchId, BatchEntry>,
+}
+
+struct BatchEntry {
+    thread_id: ThreadId,
+    num_transactions: usize,
+    total_cus: u64,
+}
+
+impl InFlightTracker {
+    pub fn new(num_threads: usize) -> Self {
+        Self {
+            num_in_flight_per_thread: vec![0; num_threads],
+            cus_in_flight_per_thread: vec![0; num_threads],
+            batches: HashMap::new(),
+        }
+    }
+
+    /// Returns the number of transactions that are in flight for each thread.
+    pub fn num_in_flight_per_thread(&self) -> &[usize] {
+        &self.num_in_flight_per_thread
+    }
+
+    /// Returns the number of cus that are in flight for each thread.
+    pub fn cus_in_flight_per_thread(&self) -> &[u64] {
+        &self.cus_in_flight_per_thread
+    }
+
+    /// Add a new batch with given `batch_id` and `num_transactions` to the
+    /// thread with given `thread_id`.
+    ///
+    /// # Panics
+    /// Panics if the batch id is already being tracked.
+    pub fn track_batch(
+        &mut self,
+        batch_id: TransactionBatchId,
+        num_transactions: usize,
+        total_cus: u64,
+        thread_id: ThreadId,
+    ) {
+        self.num_in_flight_per_thread[thread_id] += num_transactions;
+        self.cus_in_flight_per_thread[thread_id] += total_cus;
+        if self
+            .batches
+            .insert(
+                batch_id,
+                BatchEntry {
+                    thread_id,
+                    num_transactions,
+                    total_cus,
+                },
+            )
+            .is_some()
+        {
+            panic!("batch id {batch_id} is already being tracked");
```

Comment:
Interesting to panic from here. 

My thoughts are: 
1. this Tracker is a utility - trying to help making load balance decision. It does not indicate nor dictates main process' success. 
2. therefore error here should not be "unrecoverable", 
3. so it should not panic

But if a batch is being tracked, implies be processed, more than once. That is a panicking situation. 

So I'd think the check batch is not duplicated should be outside of this Tracker, rather in main logic part (maybe in Scheduler process?), and panic there sounds better to me.

Reply:
It's a logical misuse to re-use a batch id. The controlling schedule wouldn't do anything other than panic, so I just kept the error internal. Given that, maybe it makes more sense to provide an interface that looks like this:

```rust
pub fn track_new_batch(&mut self, num_transactions: usize, total_cus: u64, thread_id: ThreadId) -> TransactionBatchId;

pub fn complete_batch(&mut self, batch_id: TransactionBatchId) -> ThreadId; 
```

This way, batch id generation is internal to the tracker, we are guaranteeing batch ids don't get re-used.

guarantee is probably too strong a word, but it's exceedingly unlikely we'll re-use a batch id (would need u64::MAX batches to be tracked, and something to get stuck).

Reply:
Making Tracker to be responsible for BatchId generation, and tracking, changing it for better 😄

Reply:
Done: bb6e502db5